### PR TITLE
User story issue#32

### LINKED ISSUE #32: As a user I want to be able to apply for jobs on the system so that I can increase my chances of getting interviews.
**Story Point( Effort):**   3

**Priority :** 2 (due second sprint)

Job application Page will have job postings that the user will click on to view the full job posting (this would be a new page). They can then click on an apply button at the bottom of the job posting.
This user story has both a front end and back end that must be done.
Front end is concentrated in 2 pages: A page with a list of job postings and a page for a specific job posting.
Back end is mostly connecting with the database to show the up to date jobs posting.

### LINKED ISSUE #79: As a user I want to view jobs on the system so that I can apply for jobs that i am interrested in
USP | 5
Priority | 2
Description | The system will have a list of jobs that they can apply to. The user will be view the jobs.

### PR DESCRIPTION
**Pull request by DasDeveloper**
Added the frontend for job listing and for job application.

Added the backend for job listing and for job application.

Added tests for the frontend.

Added tests for the backend.

Added jobRoutes.

### CODE REVIEW

File: src/client/src/components/JobListing.jsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                
+                    <div key = {job._id} className="jobPost">
+                        
+                        <p> Job id: {job.job_id}</p>
+                        <p>Title: {job.title}</p>
+                        <p>Description: {job.description}</p>
+                        <p>Salary: {job.salary}</p>
+                        <p>Company: {job.company}</p>
+                        <p>Category: {job.category}</p>
+                        <Link to = {`String text ${job.job_id}`} state = {{jobState:job}}>
```
Comment (Aboudik1452) [inline src/client/src/components/JobListing.jsx:42]: Is `String text` necessary? It is displayed on the address bar
image

File: src/client/src/css/jobListing.css [hunk 2]
```diff
@@ -0,0 +1,89 @@
+
+.Container{
+
+    display: flex;
```
Comment (Aboudik1452) [inline src/client/src/css/jobListing.css:4]: While the page is responsive, it looks messy on a smaller device. Open Inspect on Chrome/Firefox to check this, and toggle Device Preview.

image

File: src/client/src/test/JobPosting.test.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import JobPosting from "../components/JobListing";
+import '@testing-library/jest-dom';
+import React from "react";
+import { render, screen, fireEvent, cleanup } from "@testing-library/react";
+
+
+test("jobListing page renders properly", ()=>{
+    render(<JobPosting/>);
+
+    const fnameTextboxElement = screen.getByRole('textbok', {name:'fname'});
```
Comment (Aboudik1452) [inline src/client/src/test/JobPosting.test.js:10]: This test failed on my system for this reason:
`TestingLibraryElementError: Unable to find an accessible element with the role "textbok" and name "fname"`

Comment (Aboudik1452) [inline src/client/src/test/JobPosting.test.js:10]: This test failed on my system for this reason:
`TestingLibraryElementError: Unable to find an accessible element with the role "textbok" and name "fname"`

File: src/client/src/components/JobListing.jsx (hunk 1 above)
Comment (DasDeveloper) [reply src/client/src/components/JobListing.jsx:42]: You're right, it was a typo thanks for catching
Screen Shot 2023-02-17 at 5 33 42 PM

File: src/client/src/css/jobListing.css [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+.jobPosts .jobPost button{
+    background-color: #19718d;
+    display: block;
+    width: 100px;
+    border-radius: 20px;
+    margin: 15px;
+    text-decoration: none;
+}
```
[the remark this answers is not in the document]
Comment (DasDeveloper) [reply src/client/src/css/jobListing.css:52]: This should be fixed by changing the height selector to auto.

Screen Shot 2023-02-17 at 5 42 29 PM

File: src/server/tests/josbApplication.test.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    })
+
+    describe('POST /user/:id/jobsApplied', function() {
+        it("it successfully adds a job id to the jobsApplied attribute", async ()=>{
+
+            await request(app)
+                .post("/users/63edb27d0e77e161a004824c/jobsApplied", {
+                    userId: '63edb27d0e77e161a004824c',
+                    jobId: '2142412'
+                })
```
Comment (Aboudik1452) [inline src/server/tests/josbApplication.test.js:38]: On my system, both parameters are not being passed through correctly, I get `undefined`.

File: src/server/routes.rest [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+GET http://localhost:9000/users/63edb27d0e77e161a004824c/jobsApplied
+
+
+###
+
+POST http://localhost:9000/users/63edb27d0e77e161a004824c/jobsApplied
+
+###
+
```
Comment (Aboudik1452) [inline src/server/routes.rest:20]: What is this file used for?

File: src/server/models/user.js [hunk 7]
```diff
@@ -44,4 +44,12 @@ const UserSchema = new mongoose.Schema({
 const User = mongoose.model("Users", UserSchema);
 
 
+// User.create({
+//     firstname: "Bob",
+//     lastname: "Test",
+//     email: "bobTest@hotmail.com",
+//     password: "122121212",
+//     jobsApplied: [1234131231242]
+// })
```
Comment (Aboudik1452) [inline src/server/models/user.js:53]: We can re-use sample data as test data for running tests.
You can remove this from the file.

File: src/server/controllers/usersController.js [hunk 8]
```diff
@@ -92,13 +93,53 @@ const getUserInfo = async (req, res) => {
...
+      jobsApplied: req.body.jobId
+    }
+  }
+  // function(err, result) {
+  //   if (err) {
+  //     res.send(err);
+  //   } else {
+  //     res.send(result);
+  //   }
+  // })
```
Comment (Aboudik1452) [inline src/server/controllers/usersController.js:125]: You can remove this from the file

File: src/server/models/Job.js [hunk 9]
```diff
@@ -29,4 +29,14 @@ const jobsSchema = new mongoose.Schema({
...
 const Jobs = mongoose.model("Jobs", jobsSchema);
 
+// Jobs.create({
+//     job_id:12341312342323,
+//     description: "test2",
+//     salary: 123,
+//     company: "testCompany2",
+//     category:"Software",
+//     title: "testTitle2"
+// })
```
Comment (Aboudik1452) [inline src/server/models/Job.js:39]: You can remove this, and re-use it later for testing.

**Review by Joejoe1452 — CHANGES_REQUESTED** [review]
Good work @DasDeveloper!

Please look at the issues mentioned and check that your functionality works. It's possible that these could be due to how my environment is setup. I will await input from the other reviewers of this pull request.

File: src/server/routes.rest (hunk 6 above)
Comment (DasDeveloper) [reply src/server/routes.rest:20]: This is used for an extension. The extensions allows us to send requests just like postman. I'll put it in my gitignore file.

File: src/client/src/css/jobListing.css (hunk 2 above)
Comment (DasDeveloper) [reply src/client/src/css/jobListing.css:4]: Made the pages responsive for smaller devices.
Screen Shot 2023-02-17 at 7 06 40 PM
For job listing and job application

File: src/server/tests/josbApplication.test.js (hunk 5 above)
Comment (DasDeveloper) [reply src/server/tests/josbApplication.test.js:38]: That test passes on my end.
Screen Shot 2023-02-17 at 7 29 41 PM

**Review by imran1289-ah — APPROVED** [review]
it looks good overall. i guess we are using some sort of a predefined id to make the request to the backend ?We will probably have to change this in the future and have it fetch the id of the logged in recruiter.

File: src/server/routes/jobsRoutes.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+//     try{
+//         const job = await Job.findOne({job_id: req.params.jobId});
+//         res.send(job);
+//     }catch(err){
+//         res.status[400].json({message: "Something went wrong."})
+//     }
+  
+// })
+
```
[the remark this answers is not in the document]
Comment (DasDeveloper) [reply src/server/routes/jobsRoutes.js:31]: They were used for debugging. I'll remove them now. Thanks for the spotting.

**Comment by DasDeveloper** [comment]
> it looks good overall. i guess we are using some sort of a predefined id to make the request to the backend ?We will probably have to change this in the future and have it fetch the id of the logged in recruiter.

Yes that's correct.

**Review by Kunal22shah — APPROVED** [review]
LGTM, if you can get the tests working with ~80% coverage it'd be great

**Comment by Aboudik1452** [comment]
Good work!

**Review by Joejoe1452 — APPROVED** [review]

**Pull request merged**